### PR TITLE
Add LabelPrefix context to better guess correct input labels

### LINKED ISSUE #7646: ReferenceInput label uses the wrong resource for translation message
ReferenceInput in react-admin@4.02 traduct field in the same way of react-admin@3.19

When editing a resource posts with the following field
`<ReferenceInput source="idUser" reference="users"><AutoComplete /></ReferenceInput>`
The ReferenceInput label will be traduct with the key resources.**users**.fields.idUser in react-admin@4, while react-admin@3.19 use resources.**posts**.fields.idUser

The ReferenceField use resources.**posts**.fields.idUser in react-admin@3.19 and react-admin@4.02

### PR DESCRIPTION
**Pull request by fzaninotto**
Fixes #7646

Instead of using the `resource`, the function that builds field and input labels can use the value of a context, which is set by the forms. This allows for better defaults in forms when using `<ReferenceInput>` and `<ArrayInput>`.

The change is made BC by keeping the previous behavior when the `resource` is passed explicitly.

I also took the opportunity to fix default input labels in `<SimpleFormIterator>` children (instead of `resources.posts.fields.aothors.123.name`, they are now `resources.posts.fields.aothors.name`).

### CODE REVIEW

File: packages/ra-core/src/util/getFieldLabelTranslationArgs.ts [hunk 1]
```diff
@@ -17,23 +18,76 @@ type TranslationArguments = [string, any?];
...
+    const sourceSuffix =
+        lastPartWithDigits != null
+            ? parts.slice(lastPartWithDigits + 1).join('.')
+            : source;
+
+    return { sourceWithoutDigits, sourceSuffix };
+};
+
+// 48 and 57 are the char codes for "0" and "9", respectively
+function onlyDigits(s) {
```
Comment (djhi) [inline packages/ra-core/src/util/getFieldLabelTranslationArgs.ts:87]: Use an arrow function for consistency?

**Review by djhi — CHANGES_REQUESTED** [review]
SimpleFormIteratorItem should set a prefix as well if I understand correctly, shouldn't it?

**Comment by fzaninotto** [comment]
> SimpleFormIteratorItem should set a prefix as well if I understand correctly, shouldn't it?

No, ArrayInput doesn't change the LabelPrefix, as developers will prefer to have the entire path in the message. For instance:

```jsx
const PostEdit = () => (
   <Edit>
     <SimpleForm>
       <ArrayInput source="backlinks">
          <SimpleFormIterator>
             <TextInput source="url" />
```

`ArrayInput` will generate for the backlinks sources like `backlinks.0.url`, `backlinks.1.url`, etc. What the developers need is a constant label for each input, BUT keeping the context of the post: `resources.posts.fields.backlinks.url`. So we shouldn't try to change the prefix in ArrayInput - this is really designed for ReferenceInputs (and other inputs reacting a new ResourceContext).

File: packages/ra-core/src/form/Form.tsx [hunk 2]
```diff
@@ -38,22 +40,25 @@ import { useAugmentedForm } from './useAugmentedForm';
...
-                        onSubmit={formHandleSubmit}
-                        noValidate={noValidate}
-                        id={id}
-                        className={className}
-                    >
-                        {children}
-                    </form>
-                </FormGroupsProvider>
-            </FormProvider>
+            <LabelPrefixContextProvider prefix={`resources.${resource}.fields`}>
```
[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/ra-core/src/form/Form.tsx:48]: What should be the label prefix if we're not in the context of a resource?

[the remark this answers is not in the document]
Comment (slax57) [reply packages/ra-core/src/form/Form.tsx:48]: IMO we should not have a `<LabelPrefixContextProvider ...>` in this case. Anyways the prefix should never be `"resources..fields"`. But maybe I'm warning for nothing, maybe this case will never happen. This is why I expressed my concern as a question 😅

[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/ra-core/src/form/Form.tsx]: I added a special prefix for that case

[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/ra-core/src/form/Form.tsx:48]: So apparently there were tests in the code to make sure that in this case the translation message is indeed `resources.undefined.fields.name`. Changing this breaks tests, and maybe also existing apps. Let's not address that case with this PR.

**Pull request merged**